NLP/main/extract_display_quotes.py

```diff
@@ -77,164 +80,87 @@ def __init__(self):
-        def _cb(change):
-            clear_output()
-            uploaded_success = widgets.HTML(
-            value='<b>File uploaded!</b>',
-            placeholder='',
-            description='',
-            layout=Layout(margin='0px 0px 0px 0px')
-            )
-            display(uploaded_success)
-            
-            # process the uploaded file
-            if tab.children[0].value!={}:
-                self.process_txt(tab.children[0])
-            else:
-                self.process_xls(tab.children[1])
-            
-        uploader_text.observe(_cb, names='data')
-        uploader_xls.observe(_cb, names='data')
-        
-        return tab
+            ) 
+
     
+
     
-    def nlp_preprocess(self, text):
+    def load_txt(self, value):
+        # Load individual txt file content and return a dict object, wrapped in a list so it can be merged with list of pervious file contents.
+        temp = {'text_name': value['metadata']['name'][:-4],
+                'text': codecs.decode(value['content'], encoding='utf-8')
+        }
+        return [temp]
+
+
+    def load_table(self, value, file_fmt):
+        # Load csv or xlsx file
+        if file_fmt == 'csv':
+            temp_df = pd.read_csv(io.BytesIO(value['content']))
+        if file_fmt == 'xlsx':
+            temp_df = pd.read_excel(io.BytesIO(value['content']))
+        # Check if the column text and text_name present in the table, if not, skip the current spreadsheet
+        if ('text' not in temp_df.columns) or ('text_name' not in temp_df.columns):
+            print('File {} does not contain the required header "text" and "text_name"'.format(value['metadata']['name']))
+            return []
+        # Return a list of dict objects
+        temp = temp_df[['text_name', 'text']].to_dict(orient='index').values()
+        return temp
+
+
+    def hash_gen(self, temp_df):
+        # Create column text_id by md5 hash of the text in text_df
+        temp_df['text_id'] = temp_df['text'].apply(lambda t: hashlib.md5(t.encode('utf-8')).hexdigest())
+        return temp_df
+
+
+    def nlp_preprocess(self, temp_df):
         '''
         Pre-process and create spaCy text
 
         Args:
             text: the text to be processed
         '''
-        # pre-process text
-        text = sent_tokenize(text)
-        text = ' '.join(text)
-        text = utils.preprocess_text(text)
-        
-        # apply the spaCy's tool to the text
-        doc = self.nlp(text)
-        
-        return doc
-    
-    
-    def process_txt(self, txt_upload):    
+        # pre-process text and fit it with Spacy language model into the column "spacy_text"
+        temp_df['spacy_text'] = temp_df['text']\
+            .map(sent_tokenize)\
+                .apply(lambda t: ' '.join(t))\
+                    .map(utils.preprocess_text)\
+                        .map(self.nlp)
+        return temp_df
+
+
+    def process_upload(self, deduplication=True):    
```

Comment:
is it possible to combine this into the upload section, so it process deduplication once the files are uploaded? then user should be able to preview them in the next cell. also it may be confusing when in the upload button the number is 56, but in the below only 28 uploaded. 
<img width="935" alt="upload and deduplication" src="https://user-images.githubusercontent.com/61028906/177925413-d388093a-0ff8-4aa2-bdda-14e543cfb229.png">

Reply:
Is it because 28/56 are duplications? Yes, the numbers on the upload widget are confusing, we should explain the deduplication process, and also the number shows how many files have been uploaded, and it's irrelevant to the actual documents to be analysed, as a spreadsheet can include many documents. Happy to make it part of the upload process, so that people can keep uploading from different sources without the need to execute the process cell.

Reply:
Though I had trouble observing the changes and trigger the process, don't know why, then I split them into two cells. But happy to give it a try again.